subprojects/scala/src/main/java/org/gradle/api/tasks/scala/internal/DefaultScalaCompileOptionsConfigurer.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.tasks.scala.internal;
+
+import org.gradle.api.internal.artifacts.ivyservice.ivyresolve.VersionInfo;
+import org.gradle.api.internal.artifacts.ivyservice.ivyresolve.strategy.DefaultVersionComparator;
+import org.gradle.api.internal.artifacts.ivyservice.ivyresolve.strategy.VersionParser;
+import org.gradle.api.tasks.ScalaRuntime;
+import org.gradle.api.tasks.scala.ScalaCompileOptions;
+import org.gradle.jvm.toolchain.JavaInstallationMetadata;
+
+import java.io.File;
+import java.util.Collections;
+import java.util.Set;
+
+/**
+ * Configures scala compile options to include the proper jvm target.
+ * Does not configure if Java Toolchain feature is not in use
+ * @since 7.3
+ */
+public class DefaultScalaCompileOptionsConfigurer implements ScalaCompileOptionsConfigurer {
+    private static final VersionParser versionParser = new VersionParser();
+    private static final DefaultVersionComparator comparator = new DefaultVersionComparator();
+
+    @Override
+    public void configure(ScalaCompileOptions scalaCompileOptions, JavaInstallationMetadata toolchain, Set<File> scalaClasspath) {
+        if (toolchain == null) {
+            return;
+        }
+
+        File scalaJar = ScalaRuntime.findScalaJar(scalaClasspath, "library");
+        if(scalaJar == null) {
+            return;
+        }
+
+        String scalaVersion = ScalaRuntime.getScalaVersion(scalaJar);
+        if(scalaVersion == null) {
+            return;
+        }
+
+        if(scalaCompileOptions.getAdditionalParameters() != null && scalaCompileOptions.getAdditionalParameters().stream().anyMatch(s -> s.startsWith("-target:"))) {
+            return;
+        }
+
+        Integer jvmVersion = toolchain.getLanguageVersion().asInt();
+        String targetParameter = determineTargetParameter(scalaVersion, jvmVersion);
+        if(scalaCompileOptions.getAdditionalParameters() == null) {
+            scalaCompileOptions.setAdditionalParameters(Collections.singletonList(targetParameter));
+        } else {
+            scalaCompileOptions.getAdditionalParameters().add(targetParameter);
+        }
+    }
+
+    private String determineTargetParameter(String scalaVersion, Integer jvmVersion) {
```

Comment:
There is something I do not understand in that logic, even after reading again the PR description.
* The format of the flag evolved across Scala version - that makes sense ✅ 
* But why is the jvm version _ignored_ in some branches?

My expectation of the logic here would have been around figuring out the right format but otherwise always use the version from the toolchain.
In short, using Java 16 toolchain for compiling with Scala 2.11.x should result in `-target:jvm-16` ... and potentially fail unless the user explicitly specifies that flag himself.

Reply:
Good feedback, I tried to make it a little bit opinionated based on what I learned from the scalac commands. I did the changes in https://github.com/gradle/gradle/pull/18347/commits/ad84107dba04e0ec39290488d6487e5adb2171d6 so now you should only expect `-target:jvm-1.8` vs `-target:8` instead of determining the "max" supported per minor/fix version